lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp

```diff
@@ -151,14 +186,74 @@ class LatticeValue {
   bool operator!=(const LatticeValue &other) const {
     return valueAndTag != other.valueAndTag;
   }
+  friend raw_ostream &operator<<(raw_ostream &os, const LatticeValue &dt);
 
 private:
   /// The attribute value if this is a constant and the tag for the element
   /// kind.  The attribute is always an IntegerAttr.
   llvm::PointerIntPair<Attribute, 2, Kind> valueAndTag;
 };
+
+raw_ostream &operator<<(raw_ostream &os, const LatticeValue &lattice) {
+  os << "<unknown: " << lattice.isUnknown()
+     << ", invalid: " << lattice.isInvalidValue()
+     << ", constant: " << lattice.isConstant()
+     << ", over: " << lattice.isOverdefined() << ">";
+  return os;
+}
+
+// This class defines the data structure associated with lattice values.
+// This means basically a pair of value and integer. The integer
+// represents the relative offset to the value on leaf level of type.
+class ValueAndLeafIndex {
+public:
+  ValueAndLeafIndex(Value value, unsigned leafId, bool isKnownRoot = false)
+      : valueAndFlag(value, isKnownRoot), leafId(leafId) {}
+  ValueAndLeafIndex() = default;
+  Value getValue() const { return valueAndFlag.getPointer(); }
+  unsigned getLeafIndex() const { return leafId; }
+
+  // isKnownRoot is true if we already know that the value is a root value.
+  // If this flag is true, we can skip the translation.
+  bool isKnownRoot() const { return valueAndFlag.getInt(); }
+  void setKnownRoot() { valueAndFlag.setInt(true); }
+
+private:
+  // Pair of a value and flag.
+  llvm::PointerIntPair<Value, 1, bool> valueAndFlag{};
+  unsigned leafId{};
```

Comment:
nit, but please use `= 0` isntad of {}.  You can drop the {} from valueAndFlag

Reply:
Thanks!